Spatial Polyakov Loops

vataspro opened the pull request:
Implement spatial Polyakov loops in a non-invasive manner.

This is done by overloading the Polyakov "avgPolyakovLoop" function and introducing a new SpatialPolyakovLoop Observable and corresponding module.

Has now been rebased to sync with telos-collaboration/Grid.

vataspro committed:
Generalise Polyakov loop and overload for temporal direction

vataspro committed:
Create Spatial Polyakov Observable Module

vataspro committed:
Add Spatial Polyakov Loop observable

edbennett commented on Grid/qcd/observables/spatial_polyakov_loop.h:
```diff
@@ -0,0 +1,74 @@
+
+Source file: ./lib/qcd/modules/polyakov_line.h
+
+Copyright (C) 2017
```
2025

edbennett commented on Grid/qcd/observables/spatial_polyakov_loop.h:
```diff
@@ -0,0 +1,74 @@
+
+Grid physics library, www.github.com/paboyle/Grid
+
+Source file: ./lib/qcd/modules/polyakov_line.h
```
Correct this filename

edbennett commented on Grid/qcd/observables/spatial_polyakov_loop.h:
```diff
@@ -0,0 +1,74 @@
+
+Copyright (C) 2017
+
+Author: David Preti <david.preti@csic.es>
```
Add your name here

edbennett commented on Grid/qcd/observables/spatial_polyakov_loop.h:
```diff
@@ -0,0 +1,74 @@
+
+// this is only defined for a gauge theory
+template <class Impl>
+class SpatialPolyakovLogger : public HmcObservable<typename Impl::Field> {
```
Does this need to be in a separate file, or could it be added to `polyakov_loop.h`?

vataspro replied:
I have added this to `polyakov_loop.h`

edbennett commented on Grid/qcd/utils/WilsonLoops.h:
```diff
@@ -177,25 +177,37 @@ template <class Gimpl> class WilsonLoops : public Gimpl {
-  static ComplexD avgPolyakovLoop(const GaugeField &Umu) {  //assume Nd=4
+  static ComplexD avgPolyakovLoop(const GaugeField &Umu, const int mu) {  //assume Nd=4
+    
+    // TODO: Protect against mu<0 || mu>=4
```
Generally not a good idea to submit PRs with `TODO` items still present.

vataspro replied:
I think I have implemented this correctly, printing an error message to the Grid error log and using exit(-1) in case of a bad mu value.

edbennett commented on Grid/qcd/utils/WilsonLoops.h:
```diff
@@ -177,25 +177,37 @@ template <class Gimpl> class WilsonLoops : public Gimpl {
 
   //////////////////////////////////////////////////
-  // average over all x,y,z the temporal loop
+  // average Polyakov loop in any direction (spatial or temporal)
```
This subtracts a little information from the original comment. Might be worth saying something like "average Polyakov loop in direction $\mu$ over all directions $\ne\mu$".

edbennett commented on Grid/qcd/utils/WilsonLoops.h:
```diff
@@ -177,25 +177,37 @@ template <class Gimpl> class WilsonLoops : public Gimpl {
+    // Number of sites in mu direction
+    int L = Umu.Grid()->GlobalDimensions()[mu];
+
+    Ut = peekLorentz(Umu, mu); //Select direction
```
`Ut` was previously so called because it was the gauge field $U$ in the $t$ direction; this is probably no longer a good name for it. (Annoyingly the name that would make sense is already taken—`Umu`. Have a think and see if you come up with a good name.)

edbennett commented on Grid/qcd/utils/WilsonLoops.h:
```diff
@@ -177,25 +177,37 @@ template <class Gimpl> class WilsonLoops : public Gimpl {
 
-    Ut = peekLorentz(Umu,3); //Select temporal direction
+    // Number of sites in mu direction
+    int L = Umu.Grid()->GlobalDimensions()[mu];
```
Usually I would interpret `L` as the spatial lattice volume; here it could also be temporal. Could `N_mu` be a good name here to represent the number of sites in direction $\mu$?

edbennett commented on Grid/qcd/utils/WilsonLoops.h:
```diff
@@ -177,25 +177,37 @@ template <class Gimpl> class WilsonLoops : public Gimpl {
+  }  
+
+  /////////////////////////////////////////////////
+  // overload for temporal loop
```
Add `Polyakov` into this comment for completeness.

edbennett reviewed (changes requested):
Thanks Alexi.

There is probably a slightly neater way to achieve this with templating so that the separate temporal and spatial implementations aren't needed, but it would be more invasive to the existing code, and would risk compatibility.

This strikes a nice balance of minimal changes to existing Grid code and minimal risk of creating any changes in behaviour of existing client code.

I've noted a few changes that would aid readability.

vataspro committed:
Move observable logger to Polyakov Loop file and fix docstring

vataspro committed:
remove obsolete spatial polyakov observable file

vataspro committed:
Fix names, protect against bad index values, clean docstrings

vataspro commented:
Hi Ed,

I think I have addressed your comments. I would like to bring your attention to:

- Removed the spatial_polyakov_loop.h and moved to polyakov_loop.h
- Protective code in WilsonLoops.h, line 184

Please let me know if there is anything else to fix.

edbennett commented on Grid/qcd/utils/WilsonLoops.h:
```diff
@@ -177,25 +177,43 @@ template <class Gimpl> class WilsonLoops : public Gimpl {
   //////////////////////////////////////////////////
-  static ComplexD avgPolyakovLoop(const GaugeField &Umu) {  //assume Nd=4
-    GaugeMat Ut(Umu.Grid()), P(Umu.Grid());
+  static ComplexD avgPolyakovLoop(const GaugeField &Umu, const int mu) {  //assume Nd=4
```
In principle, Grid supports `Nd` ≠ 4, and much code in Grid (even in the `Grid/qcd` directory) is written at least partially for generic `Nd`. 

So rather than _assuming_ this, it might be better to explicitly verify it.

vataspro replied:
Hi Ed, I'm happy to do this but this is actually from the original code for the [Polyakov Loop](https://github.com/paboyle/Grid/blob/develop/Grid/qcd/utils/WilsonLoops.h).

I would argue that this is two different pull requests, one that generalises the Polyakov Loop to Nd and one that implements the Spatial Polyakov loops. 

This would also be an invasive change.

vataspro replied:
Looking into this a bit more, the Polyakov Loop could actually be rewritten so as to get the average using the gSites method of the gauge field, as is done in line 123 of `Grid/qcd/utils/WIlsonLoops.h` (of the original Grid source code on the current develop branch) for the volume for the average plaquette.

I would still argue that this should be a separate pr, aimed at generalising the Polyakov Loop for `Nd`.

edbennett replied:
Thanks Alexi; I saw it was in green so assumed it was our change.

I agree that if this was in the original, then we shouldn't touch it. It could be a separate PR, if anyone ever finds themselves wanting a 10-dimensional Polyakov loop computation, but we don't need to think about it today.

edbennett reviewed (changes requested):
Thanks Alexi, this looks a lot tidier. Looking through this again, I spotted one remaining place that could benefit from an assumption in a comment being turned into an explicit check.

edbennett reviewed (approved):
I think this is ready to merge now.
